src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,493 @@
+          []         -> error "impossible"
+          (vo : kos) -> (kos, vo)
+
+    pageValueOffsets = case unStricterList pageValues of
+      [v] -> Right (offValues, fromIntegral offValues + sizeofValue32 v)
+      vs  -> Left (scanr (\v o -> o + sizeofValue16 v) offValues vs)
+
+    rawKey :: RawKey -> BB.Builder
+    rawKey (S.SerialisedKey (ByteArray ba#)) = BB.shortByteString (SBS ba#)
+
+    rawValue :: RawValue -> BB.Builder
+    rawValue EmptyRawValue                  = mempty
+    rawValue (RawValue i j (ByteArray ba#)) = shortByteStringFromTo i j (SBS ba#)
+
+-- | Copy of 'SBS.shortByteString', but with bounds (unchecked)
+{-# INLINE shortByteStringFromTo #-}
+shortByteStringFromTo :: Int -> Int -> ShortByteString -> BB.Builder
+shortByteStringFromTo = \i j sbs -> BB.builder $ shortByteStringCopyStepFromTo i j sbs
+
+-- | Copy of 'SBS.shortByteStringCopyStep' but with bounds (unchecked)
+{-# INLINE shortByteStringCopyStepFromTo #-}
+shortByteStringCopyStepFromTo ::
+  Int -> Int -> ShortByteString -> BB.BuildStep a -> BB.BuildStep a
+shortByteStringCopyStepFromTo !ip0 !ipe0 !sbs k =
+    go ip0 ipe0
+  where
+    go !ip !ipe (BB.BufferRange op ope)
+      | inpRemaining <= outRemaining = do
+          SBS.copyToPtr sbs ip op inpRemaining
+          let !br' = BB.BufferRange (op `plusPtr` inpRemaining) ope
+          k br'
+      | otherwise = do
+          SBS.copyToPtr sbs ip op outRemaining
+          let !ip' = ip + outRemaining
+          return $ BB.bufferFull 1 ope (go ip' ipe)
+      where
+        outRemaining = ope `minusPtr` op
+        inpRemaining = ipe - ip
+
+{-------------------------------------------------------------------------------
+  Accumulator for page contents
+-------------------------------------------------------------------------------}
+
+-- We might have preferred a representation for an incrementally constructed
+-- page that already places the raw bytes in a page-sized buffer. However, we do
+-- not know a priori what the offsets for most of the page components will be.
+-- These offsets depend on the number of keys\/values\/blob references, and the
+-- sizes of keys\/values, and we only find out these numbers as key-operation
+-- pairs are added incrementally. As such, we can only construct the raw page
+-- bytes once the current page has filled up.
+--
+-- TODO: an alternative representation that would put less load on the GC would
+-- be to use mutable vectors for each component. Where we currently only copy
+-- bytes once (from input to the output page), we would have to copy each byte
+-- one extra time to the intermediate mutable vectors. However, it would save
+-- having to allocate and de-allocate heap objects, like lists. We wouldn't know
+-- a priori exactly how large those vectors should be, but we know they are
+-- bounded, though special care should be taken for multi-page values.
+data PageAcc = PageAcc {
```

Comment:
This style makes sense for using a `Builder`, however if/when we switch to a low level `RawPage` style then the appropriate accumulator will probably just be a list of key/entry pairs.